### PR TITLE
Remove integration between tonic and WTFString

### PR DESCRIPTION
**Pull request by abarth**
We used to share memory between Dart strings and WTF::String objects by
way of the Dart externalized strings. That used to be important when the
DOM shared many strings between C++ and Dart. However, now that we don't
retain strings in C++ much anymore, we don't need this complexity.

This patch removes DartStringCache and the integration. It also unwinds
several cases where we were converting back and forth between
WTF::String and std::string for no reason. Now we use std::string more
consistently.

For the case of ParagraphBuilder::addText, we now take a raw const
char*, which more closely matches the API the DartVM exposes. That means
we do a single copy out of the VM and into the render tree at that
point.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

**Comment by abarth** [comment]
I've also added a commit that removes the integration between tonic and WTF::Vector.  There's no advantage to using WTF::Vector over std::vector, so I've switched the clients over to std::vector.

**Comment by abarth** [comment]
Added some more cleanup commits.  It's probably easier to review each commit separately.  I can send them in separate pull requests if you like.

**Comment by Hixie** [comment]
(haven't looked at the patch yet)

Is there a way to use the Dart strings directly in the text rendering without the complexity we used to have? Having to copy strings at all is unfortunate.

File: sky/engine/core/text/ParagraphBuilder.cpp [hunk 1]
```diff
@@ -136,7 +136,7 @@ ParagraphBuilder::~ParagraphBuilder()
 {
 }
 
-void ParagraphBuilder::pushStyle(Int32List& encoded, const String& fontFamily, double fontSize, double letterSpacing, double wordSpacing, double lineHeight)
+void ParagraphBuilder::pushStyle(Int32List& encoded, const char* fontFamily, double fontSize, double letterSpacing, double wordSpacing, double lineHeight)
```
Comment (Hixie) [inline sky/engine/core/text/ParagraphBuilder.cpp]: What if the author passes a null in the fontFamily string?

File: sky/engine/core/text/ParagraphBuilder.cpp [hunk 2]
```diff
@@ -207,11 +207,11 @@ void ParagraphBuilder::pop()
         m_currentRenderObject = m_currentRenderObject->parent();
 }
 
-void ParagraphBuilder::addText(const String& text)
+void ParagraphBuilder::addText(const char* text)
```
Comment (Hixie) [inline sky/engine/core/text/ParagraphBuilder.cpp]: ...or in the text?

**Comment by Hixie** [comment]
LGTM, modulo the following:

`char*` seems to imply null-terminated, since there's nowhere to store the size. But we absolutely want to make sure we don't handle nulls specially. (Consider e.g. an app built with flutter that shows UGC embedded with other content -- we don't want a stray null to be a way for users to hide key information from other users.)

It's also not obvious that what encoding code should expect `char*` to be using. I would hope it's UTF-8, but if it's coming from Dart it might be UTF-16? Anyway, even if we do use `char*` we should at least use a typedef to document the implicit encoding. `utf8char*` or something.

I don't know if the other string types we have have the same problems. I assume not.

**Comment by abarth** [comment]
> Is there a way to use the Dart strings directly in the text rendering without the complexity we used to have?

The trouble is that the VM likes to move data around in memory.  In order to have C++ reference the memory, you need to give it a stable address.  There are two ways to do that:

1) Copy the data out.
2) Copy the data out and tell the VM to use your stable address to back the string.

File: sky/engine/core/text/ParagraphBuilder.cpp (hunk 1 above)
Comment (abarth) [inline sky/engine/core/text/ParagraphBuilder.cpp]: The Dart code will detect that and set the mask bits appropriately.

File: sky/engine/core/text/ParagraphBuilder.cpp (hunk 2 above)
Comment (abarth) [inline sky/engine/core/text/ParagraphBuilder.cpp]: fromUTF8 will return a "null" String, which the impl() value of will be a nullptr.  Not sure what RenderText will do at that point.

**Comment by abarth** [comment]
> I don't know if the other string types we have have the same problems.

Looks like we used to handle that correctly but this patch makes it wrong.  We should use Dart_StringToUTF8 instead to extract the string, which returns the UTF8 length of the string.  I'll make that change before landing.

File: sky/engine/core/text/ParagraphBuilder.cpp (hunk 1 above)
Comment (Hixie) [inline sky/engine/core/text/ParagraphBuilder.cpp]: I meant a U+0000, as opposed to a null pointer.

We need to handle nulls and out-of-range values in general in our APIs but I figure we'll deal with that all at once at some point.

Comment (abarth) [inline sky/engine/core/text/ParagraphBuilder.cpp]: Ah, yeah.  That's busted.  I have a fix locally.  Just testing it out a bit.  This argument becomes a std::string, which can handle embedded U+0000.

**Pull request merged**